File: src/php/lamp-control-api/tests/Entity/LampEntityTest.php
<?php

namespace OpenAPIServer\Tests\Entity;

use DateTime;
use OpenAPIServer\Entity\LampEntity;
use PHPUnit\Framework\TestCase;
use Ramsey\Uuid\Uuid;

class LampEntityTest extends TestCase
{
    public function testCreateShouldGenerateNewIdAndTimestamps(): void
    {
        // Act
        $entity = LampEntity::create(true);

        // Assert
        $this->assertNotNull($entity->getId());
        $this->assertTrue($entity->getStatus());
        $this->assertInstanceOf(\DateTimeInterface::class, $entity->getCreatedAt());
        $this->assertInstanceOf(\DateTimeInterface::class, $entity->getUpdatedAt());
        $this->assertEquals($entity->getCreatedAt()->getTimestamp(), $entity->getUpdatedAt()->getTimestamp());
    }

    public function testWithUpdatedStatusShouldCreateUpdatedEntity(): void
    {
        // Arrange
        $originalEntity = LampEntity::create(false);
        usleep(1000); // Ensure different timestamp

        // Act
        $updatedEntity = $originalEntity->withUpdatedStatus(true);

        // Assert
        $this->assertTrue($originalEntity->getId()->equals($updatedEntity->getId()));
        $this->assertTrue($updatedEntity->getStatus());
        $this->assertEquals($originalEntity->getCreatedAt(), $updatedEntity->getCreatedAt());
        $this->assertGreaterThan($originalEntity->getUpdatedAt(), $updatedEntity->getUpdatedAt());
    }

    public function testWithUpdatedStatusShouldMaintainImmutability(): void
    {
        // Arrange
        $originalEntity = LampEntity::create(false);

        // Act
        $updatedEntity = $originalEntity->withUpdatedStatus(true);

        // Assert
        $this->assertFalse($originalEntity->getStatus());
        $this->assertTrue($updatedEntity->getStatus());
        $this->assertNotSame($originalEntity, $updatedEntity); // Different instances
    }

    public function testEqualsShouldReturnTrueForSameId(): void
    {
        // Arrange
        $id = Uuid::uuid4();
        $timestamp = new DateTime();
        $entity1 = new LampEntity($id, true, $timestamp, $timestamp);
        $entity2 = new LampEntity($id, false, (new DateTime())->modify('+1 day'), (new DateTime())->modify('+1 day'));

        // Act & Assert
        $this->assertTrue($entity1->equals($entity2));
    }

    public function testEqualsShouldReturnFalseForDifferentId(): void
    {
        // Arrange
        $timestamp = new DateTime();
        $entity1 = new LampEntity(Uuid::uuid4(), true, $timestamp, $timestamp);
        $entity2 = new LampEntity(Uuid::uuid4(), true, $timestamp, $timestamp);

        // Act & Assert
        $this->assertFalse($entity1->equals($entity2));
    }

    public function testToStringShouldReturnFormattedString(): void
    {
        // Arrange
        $entity = LampEntity::create(true);

        // Act
        $result = (string) $entity;

        // Assert
        $this->assertStringContainsString('LampEntity', $result);
        $this->assertStringContainsString($entity->getId()->toString(), $result);
        $this->assertStringContainsString('true', $result);
    }

    public function testDomainEntitySeparationShouldBeIndependentOfApiModels(): void
    {
        // This test ensures that LampEntity doesn't depend on any API model classes
        // Arrange
        $entity = LampEntity::create(true);

        // Assert - should be able to create and manipulate without API dependencies
        $reflection = new \ReflectionClass($entity);
        $apiModelNamespace = 'OpenAPIServer\Model';

        // Verify no direct dependency on API model classes in namespace
        $this->assertStringNotContainsString($apiModelNamespace, $reflection->getNamespaceName());
        $this->assertNotNull((string) $entity); // Basic functionality works
    }
}

File: src/kotlin/src/test/kotlin/com/lampcontrol/entity/LampEntityTest.kt
package com.lampcontrol.entity

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class LampEntityTest {
    @Test
    fun `should create new lamp entity with generated ID and timestamps`() {
        // When
        val entity = LampEntity.create(true)

        // Then
        assertNotNull(entity.id)
        assertEquals(true, entity.status)
        assertNotNull(entity.createdAt)
        assertNotNull(entity.updatedAt)
        assertEquals(entity.createdAt, entity.updatedAt)
    }

    @Test
    fun `should create updated entity with new status and timestamp`() {
        // Given
        val originalEntity = LampEntity.create(false)
        Thread.sleep(10) // Ensure different timestamp

        // When
        val updatedEntity = originalEntity.withUpdatedStatus(true)

        // Then
        assertEquals(originalEntity.id, updatedEntity.id)
        assertEquals(true, updatedEntity.status)
        assertEquals(originalEntity.createdAt, updatedEntity.createdAt)
        assertNotEquals(originalEntity.updatedAt, updatedEntity.updatedAt)
    }

    @Test
    fun `should maintain immutability when updating status`() {
        // Given
        val originalEntity = LampEntity.create(false)

        // When
        val updatedEntity = originalEntity.withUpdatedStatus(true)

        // Then
        assertEquals(false, originalEntity.status)
        assertEquals(true, updatedEntity.status)
        assertNotEquals(originalEntity, updatedEntity)
    }

    @Test
    fun `should verify domain entity separation from API models`() {
        // This test ensures that LampEntity doesn't depend on any API model classes
        // Given
        val entity = LampEntity.create(true)

        // Then - should be able to create and manipulate without API dependencies
        val entityClass = entity::class.java
        val apiModelPackage = "com.lampcontrol.api.models"

        // Verify no direct dependency on API model classes
        assertFalse(entityClass.name.contains(apiModelPackage))
        assertNotNull(entity.toString()) // Basic functionality works
    }
}
